commit 2d77e3a27b7b211f303fcc3a813a044838dad970
Author: Steve Dickson <steved@redhat.com>
Date:   Sat Jan 19 07:59:26 2008 -0500

    Fix bug when both crossmnt and fsid are set.
    
    When exported a filesystems with option inherited (by the crossmnt
    option) from a higherlevel filesystem, ignore filesystem specific
    options like FSID and explicit UUID.
    
    Signed-off-by: NeilBrown <neilb@suse.de>
    Signed-off-by: Steve Dickson <steved@redhat.com>

diff -up nfs-utils-1.1.1/utils/mountd/cache.c.orig nfs-utils-1.1.1/utils/mountd/cache.c
--- nfs-utils-1.1.1/utils/mountd/cache.c.orig	2008-01-22 14:15:12.000000000 -0500
+++ nfs-utils-1.1.1/utils/mountd/cache.c	2008-01-22 14:16:48.000000000 -0500
@@ -571,20 +571,25 @@ static int dump_to_cache(FILE *f, char *
 	qword_print(f, path);
 	qword_printint(f, time(0)+30*60);
 	if (exp) {
-		qword_printint(f, exp->e_flags);
+		int different_fs = strcmp(path, exp->e_path) != 0;
+		
+		if (different_fs)
+			qword_printint(f, exp->e_flags & ~NFSEXP_FSID);
+		else
+			qword_printint(f, exp->e_flags);
 		qword_printint(f, exp->e_anonuid);
 		qword_printint(f, exp->e_anongid);
 		qword_printint(f, exp->e_fsid);
 		write_fsloc(f, exp, path);
 		write_secinfo(f, exp);
 #if USE_BLKID
- 		if (exp->e_uuid == NULL) {
+ 		if (exp->e_uuid == NULL || different_fs) {
  			char u[16];
  			if (get_uuid(path, NULL, 16, u)) {
  				qword_print(f, "uuid");
  				qword_printhex(f, u, 16);
  			}
- 		} else if (exp->e_uuid) {
+ 		} else {
  			qword_print(f, "uuid");
  			qword_printhex(f, exp->e_uuid, 16);
  		}
